gon/contour.py

```diff
@@ -0,0 +1,148 @@
+
+    @classmethod
+    def from_raw(cls, raw: RawContour) -> 'Contour':
+        return cls(Point.from_raw(raw_vertex) for raw_vertex in raw)
+
+    @cached.property_
+    def normalized(self) -> 'Contour':
+        vertices = self._vertices
+        min_index = min(range(len(vertices)),
+                        key=vertices.__getitem__)
+        return Contour(vertices[min_index:] + vertices[:min_index])
+
+    @cached.property_
+    def orientation(self) -> 'Orientation':
+        vertices = self.normalized.vertices
+        return to_orientation(vertices[0], vertices[-1], vertices[1])
+
+    def to_clockwise(self) -> 'Contour':
+        return (self
+                if self.orientation is Orientation.CLOCKWISE
+                else self.reverse())
+
+    def to_counterclockwise(self) -> 'Contour':
+        return (self
+                if self.orientation is Orientation.COUNTERCLOCKWISE
+                else self.reverse())
+
+    def reverse(self) -> 'Contour':
+        vertices = self._vertices
+        return Contour(vertices[:1] + vertices[:0:-1])
+
+    def validate(self) -> None:
+        if len(self._vertices) < MIN_VERTICES_COUNT:
+            raise ValueError('Contour should have '
+                             'at least {expected} vertices, '
+                             'but found {actual}.'
+                             .format(expected=MIN_VERTICES_COUNT,
+                                     actual=len(self._vertices)))
+        if any(orientation is Orientation.COLLINEAR
+               for orientation in _vertices_to_orientations(self._vertices)):
+            raise ValueError('Consecutive vertices triplets '
+                             'should not be on the same line.')
+        if _vertices_form_self_intersecting_contour(self._vertices):
+            raise ValueError('Contour should not be self-intersecting.')
+
+
+def _vertices_to_orientations(vertices: Vertices) -> Iterable[Orientation]:
+    vertices_count = len(vertices)
+    return (to_orientation(vertices[index - 1], vertices[index],
+                           vertices[(index + 1) % vertices_count])
+            for index in range(vertices_count))
+
+
+def _vertices_form_strict_contour(vertices: Vertices) -> bool:
+    return all(orientation is not Orientation.COLLINEAR
+               for orientation in _vertices_to_orientations(vertices))
+
+
+def _vertices_form_self_intersecting_contour(vertices: Vertices) -> bool:
+    return edges_intersect([vertex.raw() for vertex in vertices])
+
+
+def forms_convex_polygon(contour: Contour) -> bool:
+    if len(contour.vertices) == 3:
+        return True
+    orientations = iter(_vertices_to_orientations(contour.vertices))
+    base_orientation = next(orientations)
+    # orientation change means
+    # that internal angle is greater than 180 degrees
+    return all(orientation is base_orientation for orientation in orientations)
+
+
+def _to_first_vertex(contour: Contour) -> Point:
+    return contour.vertices[0]
```

Comment:
Why do we need this?

Reply:
to normalize polygon by sorting holes based on the first vertex